File: kevoree-tools/org.kevoree.tools.javase.framework/src/main/scala/org/kevoree/framework/ChannelTypeFragmentThread.scala
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kevoree.framework

import internal.MethodAnnotationResolver
import message._
import org.slf4j.LoggerFactory
import java.util.HashMap
import org.kevoree.ContainerRoot
import port.PausablePortThreadPoolExecutor
import java.util.concurrent.Callable
import org.kevoree.annotation.{LocalBindingUpdated, RemoteBindingUpdated}

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 18/08/12
 * Time: 10:24
 */

trait ChannelTypeFragmentThread extends KevoreeChannelFragment with KInstance with ChannelFragment {

  val kevoree_internal_logger = LoggerFactory.getLogger(this.getClass)

  var eventHandler: event.MonitorEventHandler = null
  var pool = PausablePortThreadPoolExecutor.newPausableThreadPool(1)
  val resolver : MethodAnnotationResolver = new MethodAnnotationResolver(this.getClass);

  override def remoteDispatch(msg: Message): Object = {
    if (msg.inOut.booleanValue) {
      (this !? msg).asInstanceOf[Object]
    } else {
      this ! msg
      null
    }
  }

  private val portsBinded: HashMap[String, KevoreePort] = new HashMap[String, KevoreePort]()
  private var fragementBinded: scala.collection.immutable.HashMap[String, KevoreeChannelFragment] = scala.collection.immutable.HashMap[String, KevoreeChannelFragment]()

  def isStarted: Boolean = ct_started

  override def getBindedPorts(): java.util.List[KevoreePort] = {
    import scala.collection.JavaConversions._
    portsBinded.values.toList
  }

  //OVERRIDE BY FACTORY
  override def getOtherFragments(): java.util.List[KevoreeChannelFragment] = {
    import scala.collection.JavaConversions._
    fragementBinded.values.toList
  }

  override def forward(delegate: KevoreeChannelFragment, inmsg: Message): Object = {
    val msg = inmsg.clone
    msg.setDestChannelName(delegate.getName)
    msg.setDestNodeName(delegate.getNodeName)
    if (msg.inOut.booleanValue) {
      (delegate !? msg).asInstanceOf[Object]
    } else {
      (delegate ! msg)
      null
    }
  }


  override def forward(delegate: KevoreePort, inmsg: Message): Object = {
    val msg = inmsg.clone
    if (msg.inOut.booleanValue) {
      (delegate !? msg.getContent).asInstanceOf[Object]
    } else {
      (delegate ! msg.getContent)
      null
    }
  }

  private def createPortKey(a: Any): String = {
    a match {
      case msg: PortBindMessage => msg.getNodeName + "-" + msg.getComponentName + "-" + msg.getPortName
      case msg: PortUnbindMessage => msg.getNodeName + "-" + msg.getComponentName + "-" + msg.getPortName
      case msg: FragmentBindMessage => msg.getChannelName + "-" + msg.getFragmentNodeName
      case msg: FragmentUnbindMessage => msg.getChannelName + "-" + msg.getFragmentNodeName
      case _ => ""
    }
  }


  private var ct_started: Boolean = false

  def kInstanceStart(tmodel: ContainerRoot): Boolean = {
    if (!ct_started) {
      try {
        getModelService.asInstanceOf[ModelHandlerServiceProxy].setTempModel(tmodel)
        startChannelFragment
        getModelService.asInstanceOf[ModelHandlerServiceProxy].unsetTempModel()
        ct_started = true
        true
      } catch {
        case _@e => {
          kevoree_internal_logger.error("Kevoree Channel Instance Start Error !", e)
          false
        }
      }
    } else {
      false
    }
  }

  def kInstanceStop(tmodel: ContainerRoot): Boolean = {
    if (ct_started) {
      try {
        //TODO CHECK QUEUE SIZE AND SAVE STATE
        getModelService.asInstanceOf[ModelHandlerServiceProxy].setTempModel(tmodel)
        stopChannelFragment
        getModelService.asInstanceOf[ModelHandlerServiceProxy].unsetTempModel()
        ct_started = false
        true
      } catch {
        case _@e => {
          kevoree_internal_logger.error("Kevoree Channel Instance Stop Error !", e)
          false
        }
      }
    } else {
      false
    }
  }

  def getDictionary: java.util.HashMap[String, Object]

  def getName: String

  def kUpdateDictionary(d: java.util.HashMap[String, AnyRef], cmodel: ContainerRoot): java.util.HashMap[String, AnyRef] = {
    try {
      import scala.collection.JavaConversions._
      val previousDictionary = getDictionary.clone()
      d.keySet.foreach {
        v =>
          getDictionary.put(v, d.get(v))
      }
      if (ct_started) {
        getModelService.asInstanceOf[ModelHandlerServiceProxy].setTempModel(cmodel)
        updateChannelFragment
        getModelService.asInstanceOf[ModelHandlerServiceProxy].unsetTempModel()
      }
      previousDictionary.asInstanceOf[java.util.HashMap[String, AnyRef]]
    } catch {
      case _@e => {
        kevoree_internal_logger.error("Kevoree Channel Instance Update Error !", e)
        null
      }
    }
  }


  def startC {
    pool = PausablePortThreadPoolExecutor.newPausableThreadPool(1)
    //start()
  }

  def stopC {
    if (pool != null) {
      pool.shutdownNow()
      pool = null
    }
  }

  case class AsyncCall(o: Any) extends Runnable {
    def run() {
      if (eventHandler != null) {
        eventHandler.triggerEvent(event.MonitorEvent(classOf[ChannelFragment], getName))
      }
      o match {
        case msg: Message => {
          dispatch(msg)
        }
        case msg: Object => {
          val msg2 = new Message
          msg2.setInOut(false)
          msg2.setContent(msg)
          dispatch(msg2)
        }
      }
    }
  }

  case class SyncCall(o: Any) extends Callable[Any] {
    def call(): Any = {
      if (eventHandler != null) {
        eventHandler.triggerEvent(event.MonitorEvent(classOf[ChannelFragment], getName))
      }
      o match {
        case msg: Message => {
          if (msg.inOut.booleanValue) {
            dispatch(msg)
          } else {
            dispatch(msg)
          }
        }
        case msg: MethodCallMessage => {
          val msg2 = new Message
          msg2.setInOut(true)
          msg2.setContent(msg)
          dispatch(msg2)
        }
      }
    }
  }

  /* LifeCycle Method */
  def startChannelFragment(): Unit = {}

  def stopChannelFragment(): Unit = {}

  def updateChannelFragment: Unit = {}

  def processAdminMsg(o: Any): Boolean = {
    pool.pause()
    val res = o match {
      case msg: FragmentBindMessage => {
        kevoree_internal_logger.debug("FragmentBindMessage=>" + createPortKey(msg))
        val sender = this.createSender(msg.getFragmentNodeName, msg.getChannelName)
        val proxy = new KevoreeChannelFragmentThreadProxy(msg.getFragmentNodeName, msg.getChannelName)
        proxy.setChannelSender(sender)
        fragementBinded += ((createPortKey(msg), proxy))
        proxy.startC
        val met = resolver.resolve(classOf[RemoteBindingUpdated])
        if (met != null) {
          met.invoke(this)
        }
        true
      }
      case msg: FragmentUnbindMessage => {
        kevoree_internal_logger.debug("Try to unbind channel " + getName)
        val actorPort: Option[KevoreeChannelFragment] = fragementBinded.get(createPortKey(msg))
        if (actorPort.isDefined) {
          actorPort.get.stopC
          fragementBinded = fragementBinded.filter(p => p._1 != (createPortKey(msg)))
          val met = resolver.resolve(classOf[RemoteBindingUpdated])
          if (met != null) {
            met.invoke(this)
          }
          true
        } else {
          kevoree_internal_logger.debug("Can't unbind Fragment " + createPortKey(msg))
          false
        }
      }
      case msg: PortBindMessage => {
        portsBinded.put(createPortKey(msg), msg.getProxy)
        val met = resolver.resolve(classOf[LocalBindingUpdated])
        if (met != null) {
          met.invoke(this)
        }
        true
      }
      case msg: PortUnbindMessage => {
        portsBinded.remove(createPortKey(msg))
        val met = resolver.resolve(classOf[LocalBindingUpdated])
        if (met != null) {
          met.invoke(this)
        }
        true
      }
    }
    pool.resume()
    res
  }

  def !(o: Any) {
    pool.submit(AsyncCall(o))
  }

  def !?(o: Any): Any = {
    pool.submit(SyncCall(o)).get()
  }

}

File: kevoree-tools/org.kevoree.tools.javase.framework/src/main/java/org/kevoree/framework/ChannelTypeFragmentThread.kt
package org.kevoree.framework;

import java.util.HashMap
import java.util.concurrent.Callable
import org.kevoree.ContainerRoot
import org.kevoree.annotation.LocalBindingUpdated
import org.kevoree.annotation.RemoteBindingUpdated
import org.kevoree.api.service.core.handler.KevoreeModelHandlerService
import org.kevoree.framework.internal.MethodAnnotationResolver
import org.kevoree.framework.message.FragmentBindMessage
import org.kevoree.framework.message.FragmentUnbindMessage
import org.kevoree.framework.message.Message
import org.kevoree.framework.message.PortBindMessage
import org.kevoree.framework.message.PortUnbindMessage
import org.kevoree.framework.port.PausablePortThreadPoolExecutor
import org.slf4j.LoggerFactory
import org.kevoree.framework.internal.FieldAnnotationResolver
import org.kevoree.annotation.KevoreeInject
import java.lang.reflect.Modifier
import org.kevoree.api.Bootstraper
import org.kevoree.api.service.core.script.KevScriptEngineFactory

class ChannelTypeFragmentThread(val target: AbstractChannelFragment, val _nodeName: String, val _name: String, val modelService: KevoreeModelHandlerService,val bootService:Bootstraper,val kevsEngine : KevScriptEngineFactory): KevoreeChannelFragment, KInstance, ChannelFragment {

    public fun initChannel(){
        target.delegate = this
    }

    public override fun dispatch(msg: Message?): Any? {
        if(msg!!.getInOut()){
            return sendWait(msg)
        } else {
            send(msg)
            return null
        }
    }
    public override fun createSender(remoteNodeName: String?, remoteChannelName: String?): ChannelFragmentSender? {
        return target.createSender(remoteNodeName, remoteChannelName)
    }
    public override fun getNodeName(): String {
        return _nodeName
    }
    public override fun getName(): String {
        return _name
    }

    val kevoree_internal_logger = LoggerFactory.getLogger(this.javaClass)!!
    var pool: PausablePortThreadPoolExecutor? = null
    val portsBinded: MutableMap<String, KevoreePort> = HashMap<String, KevoreePort>()
    val fragementBinded: MutableMap<String, KevoreeChannelFragment> = HashMap<String, KevoreeChannelFragment>()
    var isStarted: Boolean = false
    val resolver: MethodAnnotationResolver = MethodAnnotationResolver(target.javaClass)
    private val fieldResolver = FieldAnnotationResolver(target.javaClass);


    override fun kInstanceStart(tmodel: ContainerRoot): Boolean {
        if (!isStarted) {
            try {

                val modelServiceFields = fieldResolver.resolve(javaClass<KevoreeInject>(), javaClass<KevoreeModelHandlerService>())!!
                for(mserv in modelServiceFields){
                    if(Modifier.isPrivate(mserv.getModifiers())){
                        mserv.setAccessible(true);
                    }
                    mserv.set(target, modelService)
                }
                val bootServiceField = fieldResolver.resolve(javaClass<KevoreeInject>(), javaClass<Bootstraper>())!!
                for(loopField in bootServiceField){
                    if(Modifier.isPrivate(loopField.getModifiers())){
                        loopField.setAccessible(true);
                    }
                    loopField.set(target, bootService)
                }
                val kevSField = fieldResolver.resolve(javaClass<KevoreeInject>(), javaClass<KevScriptEngineFactory>())!!
                for(loopField in kevSField){
                    if(Modifier.isPrivate(loopField.getModifiers())){
                        loopField.setAccessible(true);
                    }
                    loopField.set(target, kevsEngine)
                }

                target.setName(_name)
                target.setNodeName(_nodeName)


                (target.getModelService() as ModelHandlerServiceProxy).setTempModel(tmodel)
                val met = resolver.resolve(javaClass<org.kevoree.annotation.Start>())
                met?.invoke(target)
                (target.getModelService() as ModelHandlerServiceProxy).unsetTempModel()
                isStarted = true
                pool = PausablePortThreadPoolExecutor.newPausableThreadPool(1)
                return true
            } catch(e: Exception) {
                kevoree_internal_logger.error("Kevoree Channel Instance Start Error !", e)
                return false
            }
        } else {
            return false
        }
    }

    override fun kInstanceStop(tmodel: ContainerRoot): Boolean {
        if (isStarted) {
            try {
                if (pool != null) {
                    pool!!.shutdownNow()
                    pool = null
                }
                //TODO CHECK QUEUE SIZE AND SAVE STATE
                (target.getModelService() as ModelHandlerServiceProxy).setTempModel(tmodel)
                val met = resolver.resolve(javaClass<org.kevoree.annotation.Stop>())
                met?.invoke(target)
                (target.getModelService() as ModelHandlerServiceProxy).unsetTempModel()
                isStarted = false
                return true
            } catch(e: Exception) {
                kevoree_internal_logger.error("Kevoree Channel Instance Stop Error !", e)
                return false
            }
        } else {
            return false
        }
    }

    public override fun kUpdateDictionary(d: Map<String, Any>, cmodel: ContainerRoot): Map<String, Any>? {
        try {
            val previousDictionary = target.getDictionary()!!.clone()
            for(v in d.keySet()) {
                target.getDictionary()!!.put(v, d.get(v)!!)
            }
            if (isStarted) {
                (target.getModelService() as ModelHandlerServiceProxy).setTempModel(cmodel)
                val met = resolver.resolve(javaClass<org.kevoree.annotation.Update>())
                met?.invoke(target)
                (target.getModelService() as ModelHandlerServiceProxy).unsetTempModel()
            }
            return previousDictionary as Map<String, Any>?
        } catch(e: Exception) {
            kevoree_internal_logger.error("Kevoree Group Instance Update Error !", e)
            return null
        }
    }

    override fun send(o: Any?) {
        pool!!.submit(AsyncCall(o))
    }

    override fun sendWait(o: Any?): Any? {
        return pool!!.submit(SyncCall(o)).get()
    }

    inner class AsyncCall(val o: Any?): Runnable {
        override fun run() {
            when(o) {
                is Message -> {
                    target.dispatch(o)
                }
                else -> {
                    val msg2 = Message()
                    msg2.setInOut(false)
                    msg2.setContent(o!!)
                    target.dispatch(msg2)
                }
            }
        }
    }

    inner class SyncCall(val o: Any?): Callable<Any> {
        override fun call(): Any? {
            when(o) {
                is Message -> {
                    return dispatch(o)
                }
                is MethodCallMessage -> {
                    val msg2 = Message()
                    msg2.setInOut(true)
                    msg2.setContent(o)
                    return target.dispatch(msg2)
                }
                else -> {
                    val msg2 = Message()
                    msg2.setInOut(true)
                    msg2.setContent(o!!)
                    return target.dispatch(msg2)
                }
            }
        }
    }

    public fun forward(delegate: KevoreeChannelFragment?, inmsg: Message?): Any? {
        val msg = inmsg!!.clone()
        msg.setDestChannelName(delegate!!.getName())
        msg.setDestNodeName(delegate!!.getNodeName())
        if (msg.getInOut()) {
            return delegate.sendWait(msg)
        } else {
            delegate.send(msg)
            return null
        }
    }

    public fun forward(delegate: KevoreePort?, inmsg: Message?): Any? {
        val msg = inmsg!!.clone()
        msg.setDestChannelName(delegate!!.getName()!!)
        if (msg.getInOut()) {
            return delegate.sendWait(msg.getContent())
        } else {
            delegate.send(msg.getContent())
            return null
        }
    }

    override fun processAdminMsg(o: Any): Boolean {
        pool?.pause()
        val res = when(o) {
            is FragmentBindMessage -> {
                kevoree_internal_logger.debug("FragmentBindMessage=>" + createPortKey(o))
                val sender = this.createSender((o as FragmentBindMessage).fragmentNodeName, (o as FragmentBindMessage).channelName)
                val proxy = KevoreeChannelFragmentThreadProxy((o as FragmentBindMessage).fragmentNodeName, (o as FragmentBindMessage).channelName)
                proxy.channelSender = sender
                fragementBinded.put(createPortKey(o), proxy)
                //proxy.startC()
                val met = resolver.resolve(javaClass<RemoteBindingUpdated>())
                if (met != null) {
                    met.invoke(target)
                }
                true
            }
            is FragmentUnbindMessage -> {
                kevoree_internal_logger.debug("Try to unbind channel ")
                val actorPort: KevoreeChannelFragment? = fragementBinded.get(createPortKey(o))
                if (actorPort != null) {
                    //actorPort.stopC()
                    fragementBinded.remove(createPortKey(o))
                    val met = resolver.resolve(javaClass<RemoteBindingUpdated>())
                    if (met != null) {
                        met.invoke(target)
                    }
                    true
                } else {
                    kevoree_internal_logger.debug("Can't unbind Fragment " + createPortKey(o))
                    false
                }
            }
            is PortBindMessage -> {
                portsBinded.put(createPortKey(o), (o as PortBindMessage).proxy)
                val met = resolver.resolve(javaClass<LocalBindingUpdated>())
                if (met != null) {
                    met.invoke(target)
                }
                true
            }
            is PortUnbindMessage -> {
                portsBinded.remove(createPortKey(o))
                val met = resolver.resolve(javaClass<LocalBindingUpdated>())
                if (met != null) {
                    met.invoke(target)
                }
                true
            }
            else -> {
                false
            }
        }
        pool?.resume()
        return res
    }

    private fun createPortKey(a: Any): String {
        when(a) {
            is PortBindMessage -> {
                return (a as PortBindMessage).nodeName + "-" + (a as PortBindMessage).componentName + "-" + (a as PortBindMessage).portName
            }
            is PortUnbindMessage -> {
                return (a as PortUnbindMessage).nodeName + "-" + (a as PortUnbindMessage).componentName + "-" + (a as PortUnbindMessage).portName
            }
            is FragmentBindMessage -> {
                return (a as FragmentBindMessage).channelName + "-" + (a as FragmentBindMessage).fragmentNodeName
            }
            is FragmentUnbindMessage -> {
                return (a as FragmentUnbindMessage).channelName + "-" + (a as FragmentUnbindMessage).fragmentNodeName
            }
            else -> {
                return ""
            }
        }
    }


    fun getBindedPorts(): List<KevoreePort> {
        return portsBinded.values().toList()
    }

    //OVERRIDE BY FACTORY
    fun getOtherFragments(): List<KevoreeChannelFragment> {
        return fragementBinded.values().toList()
    }

}





